### PR TITLE
Billing roles

### PR DESCRIPTION
**Pull request by dvoet**
GAWB-814 and GAWB-881

added a ROLE column to the PROJECT_USER table
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Make sure liquibase is updated if appropriate
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] **Submitter**: If PR includes new or changed db queries, include the explain plans in the description
- [ ] **Submitter**: Update FISMA documentation if changes to:
  - Authentication
  - Authorization
  - Encryption
  - Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell  that the PR exists if he wants to look at it **(apply requires_doge label)**
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- Review cycle:
  - LR reviews
  - Rest of team may comment on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH
  - Submitter updates documentation as needed
  - Submitter **reassigns to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to develop 
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala [hunk 1]
```diff
@@ -51,10 +52,12 @@ object UserService {
...
-  case class ListBillingProjectsForUser(userEmail: RawlsUserEmail) extends UserServiceMessage
-  case class CreateBillingProject(projectName: RawlsBillingProjectName) extends UserServiceMessage
-  case class DeleteBillingProject(projectName: RawlsBillingProjectName) extends UserServiceMessage
-  case class AddUserToBillingProject(projectName: RawlsBillingProjectName, userEmail: RawlsUserEmail) extends UserServiceMessage
+  case class AdminListBillingProjectsForUser(userEmail: RawlsUserEmail) extends UserServiceMessage
+  case class AdminCreateBillingProject(projectName: RawlsBillingProjectName) extends UserServiceMessage
+  case class AdminDeleteBillingProject(projectName: RawlsBillingProjectName) extends UserServiceMessage
+  case class AdminAddUserToBillingProject(projectName: RawlsBillingProjectName, userEmail: RawlsUserEmail, role: ProjectRole) extends UserServiceMessage
+  case class AdminRemoveUserFromBillingProject(projectName: RawlsBillingProjectName, userEmail: RawlsUserEmail) extends UserServiceMessage
+  case class AddUserToBillingProject(projectName: RawlsBillingProjectName, userEmail: RawlsUserEmail, role: ProjectRole) extends UserServiceMessage
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:64]: make clear which messages call admin functions

File: src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala [hunk 2]
```diff
@@ -114,6 +117,16 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
     case AdminDeleteAllRefreshTokens => asAdmin { deleteAllRefreshTokens() } pipeTo sender
   }
 
+  def asOwner(projectName: RawlsBillingProjectName)(op: => Future[PerRequestMessage]): Future[PerRequestMessage] = {
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala]: Can we rename either this or `asAdmin`? Firecloud admin is not the same kind of role as project owner/user. Maybe `asFCAdmin` and/or `asProjectOwner`?

**Comment by helgridly** [comment]
Does "If PR includes new or changed db queries, include the explain plans in the description" apply here?

File: src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala (hunk 1 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:64]: is this a note to yourself?

File: src/main/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiService.scala [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+trait BillingApiService extends HttpService with PerRequestCreator with UserInfoDirectives {
+  implicit val executionContext: ExecutionContext
+
+  import org.broadinstitute.dsde.rawls.model.UserJsonSupport._
+  import spray.httpx.SprayJsonSupport._
+
+  val userServiceConstructor: UserInfo => UserService
+
+  val billingRoutes = requireUserInfo() { userInfo =>
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiService.scala:23]: (how) is this different to the endpoints at user/billing and admin/billing? Should they live here too?

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiServiceSpec.scala [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+import org.broadinstitute.dsde.rawls.model._
+import org.broadinstitute.dsde.rawls.openam.MockUserInfoDirectives
+import org.broadinstitute.dsde.rawls.user.UserService
+import spray.http.StatusCodes
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, ExecutionContext}
+
+/**
+ * Created by tsharpe on 9/28/15.
```
Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiServiceSpec.scala]: ?!

**Comment by helgridly** [comment]
review doned

File: src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala (hunk 1 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:64]: no, it is a note to you

File: src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala (hunk 2 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala]: both sound good to me

File: src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala (hunk 1 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:64]: oh i see. yes, well done

File: src/main/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiService.scala (hunk 3 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiService.scala:23]: user/billing is the user specific billing stuff, i.e. the user's billing accounts or projects. This is for strictly project management apis. Should admin/billing belong here? Possibly

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiServiceSpec.scala (hunk 4 above)
Comment (dvoet) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiServiceSpec.scala]: copy/paste :(

**Comment by dvoet** [comment]
explain plan for new query:

```
explain select count(1) from `PROJECT_USERS` where ((`PROJECT_NAME` = 'broad-dsde-dev') and (`USER_SUBJECT_ID` = '110707025587602912910')) and (`ROLE` in ('User', 'Owner'));
+----+-------------+---------------+-------+----------------------------------+---------+---------+-------------+------+-------+
| id | select_type | table         | type  | possible_keys                    | key     | key_len | ref         | rows | Extra |
+----+-------------+---------------+-------+----------------------------------+---------+---------+-------------+------+-------+
|  1 | SIMPLE      | PROJECT_USERS | const | PRIMARY,FK_PROJECT_USERS_PROJECT | PRIMARY | 1528    | const,const |    1 | NULL  |
+----+-------------+---------------+-------+----------------------------------+---------+---------+-------------+------+-------+
```

it is a bit tough because there are only a couple hundred rows in the table but the query hits the primary key.

**Comment by dvoet** [comment]
@helgridly back to you

**Comment by helgridly** [comment]
:+1:

**Pull request merged**